keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -63,67 +63,61 @@ Items marked with (R) are required *prior to targeting to a milestone / release*
 [#18787](https://github.com/kubernetes/kubernetes/issues/18787) has been around
 for a while.
 
-To secure images one currently needs to inject `AllwaysPullImages` into pod
+To secure images one currently needs to inject `Always` `ImagePullPolicy` into pod
 specs via an admission plugin. As @liggitt [notes](https://github.com/kubernetes/kubernetes/issues/18787#issuecomment-532280931)
 the `pull` does not re-pull already-pulled layers of the image, but simply
 resolves/verifies the image manifest has not changed in the registry (which
 incidentally requires authenticating to private registries, which enforces the
 image access). That means in the normal case (where the image has not changed
-since the last pull), the request size is O(kb). However, the `pull` does put
-the registry in the critical path of starting a container, since an unavailable
-registry will fail the pull image manifest check (with or without proper
-authentication.)
+since the last pull), the request size is O(kb).
+
+However, the `pull` does put the registry in the critical path of starting a container,
+since an unavailable registry will fail the pull image manifest check (with or without proper authentication.)
+
+Thus, the motivation is to allow users to ensure the kubelet requires an image pull auth check for each new set of credentials,
+regardless of whether the image is already present on the node.
 
 ### Goals
 
-Modify the current pullIfNotPresent policy management enforced by `kubelet` to
-ensure the images pulled with a secret by `kubelet` since boot. During the
-EnsureImagesExist step `kubelet` will require authentication of present images
-pulled with auth since boot.
+Modify the current behavior of images with an `IfNotPresent` and `Never` `ImagePullPolicy` enforced by the kubelet to
+ensure the images pulled with a secret by the kubelet are authenticated by the CRI implementation. During the
+EnsureImagesExist step the kubelet will require authentication of present images pulled with auth since boot.
 
 Optimize to only force re-authentication for a pod container image when the
-secret used to pull the container image is not present. IOW if an image is
-pulled with authentication for a first pod, subsequent pods that have the same
-authentication information should not need to re-authenticate.
+`ImagePullSecrets` used to pull the container image has not already been authenticated.
+IOW if an image is pulled with authentication for a first pod, subsequent pods that have the same
+authentication information should not need to re-authenticate, unless the kubelet's `pullImageSecretRecheckPeriod` has passed.
 
-Images already present at boot or loaded externally to `kubelet` or successfully
-pulled through `kubelet` with no imagePullSecret/authentication required will
+Images already present at boot or loaded externally to the kubelet or successfully
+pulled through the kubelet with no `ImagePullSecrets`/authentication required will
```

Comment:
What about images pulled directly by the runtime with creds supplied to the runtime? Is that something to mention as a non goal initially?

Reply:
this is functionally the same as preloading the image, which the kubelet must treat as a non authorized pull as it doesn't have a cache entry for it